pilot/pkg/xds/bootstrapds.go

```diff
@@ -0,0 +1,56 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xds
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	bootstrapv3 "github.com/envoyproxy/go-control-plane/envoy/config/bootstrap/v3"
+	"github.com/golang/protobuf/jsonpb"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/util"
+	"istio.io/istio/pkg/bootstrap"
+)
+
+// Bootstrap generator produces an Envoy bootstrap from node descriptors.
+type BootstrapGenerator struct {
+	Server *DiscoveryServer
+}
+
+var _ model.XdsResourceGenerator = &BootstrapGenerator{}
+
+// Generate returns a bootstrap discovery response.
+func (e *BootstrapGenerator) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource, req *model.PushRequest) (model.Resources, error) {
+	// The model.Proxy information is incomplete, re-parse the discovery request.
+	node := bootstrap.ConvertXDSNodeToNode(w.LastRequest.Node)
```

Comment:
can we get it from node.metadata.raw at least? I am not sure how reliable `w.LastRequest.Node` is going to be, seems like the wrong field to depend on even if it happens to work right now

Reply:
Can you clarify a bit more? I think `ParseMetadata` does not store the remainder of the struct into `raw` field. It does parse it as `BootstrapNodeMetadata`, and then just takes out the embedded struct (so no real performance gain from skipping extra fields). Let me add a field to store BootstrapNodeMetadata since it's technically free, and we can always track the usage of it.

Reply:
I think I need Node for locality besides Metadata. Let me store it in Proxy.

Reply:
`node.metadata.raw` is not sufficient to provide service cluster and node ID. I think the original Node is the easiest way to combine the code. Maybe just store it in `Proxy`? That would guarantee reliability.

Reply:
Yeah that seems like the best option

Reply:
Done.